lib/telemetry_sampler/vm.ex

```diff
@@ -0,0 +1,72 @@
+defmodule Telemetry.Sampler.VM do
+  @moduledoc """
+  Collection of functions dispatching `Telemetry` events with Erlang VM metrics.
+
+  See documentation for `Telemetry.Sampler` to learn how to use these functions.
+  """
+
+  @doc """
+  Dispatches events with amount of memory dynamically allocated by the VM.
+
+  A single event is dispatched for each type of memory measured. Event name is always
+  `[:vm, :memory]`. Event metadata includes only a single key, `:type`, which corresponds to the
+  type of memory measured. Event value is the amount of memory of type given in metadata allocated
+  by the VM, in bytes.
+
+  The set of memory types may vary: see documentation for `:erlang.memory/0` to learn about possible
+  values.
+  """
+  @spec memory() :: :ok
+  def memory() do
+    :erlang.memory()
+    |> Enum.each(fn {type, size} ->
+      Telemetry.execute([:vm, :memory], size, %{type: type})
```

Comment:
I am wondering if it would be better to log those as `[:vm, :memory, type]` because otherwise for things like statsD only one of the measurements will be applied, no?

The other option is to have this function reporting the sum of the memory and we have individual functions for the other memory entries. So `[:vm, :memory, :all]` and then `[:vm, :memory, :binary]`, etc, etc. It seems more granular. What do you think?

Reply:
But the StatsD metric could have a different name. `[:vm, :memory]` is an event name internal to the Erlang VM. I imagine high-level StatsD exporter API where we can construct metric name based on metadata, e.g.:

```elixir
StastD.export(%{
  [:vm, :memory] => %{template: "vm.memory.{% type %}"}
})
```

But I'm not sure, I like the second approach you proposed, too

Reply:
@arkgil I see. The reason why I find this confusing is because aggregating the values of different types across time will not give me a proper value unless I take only the latest of each type. I am not sure how flexible the tools would be to allow us to express this logic. If most tools would require us to rewrite it, then I think splitting the name apart would be better.

This is different from something like the controller+action metadata when rendering a controller+action because getting a subset of the data there is truly a subset. There is no need for grouping.

Does it make sense?

Reply:
I think I understand now. Breaking down memory by types provided by `:erlang.memory/0` doesn't really make sense, because these types are not true subsets of total memory (+ it's hard to aggregate them, as you said). I'll break it down into multiple measurements.